dev/source/docs/unit-tests.rst

```diff
@@ -0,0 +1,200 @@
+.. _ardupilot-unit-tests:
+
+====================
+ArduPilot Unit Tests
+====================
+
+.. image:: ../images/autotest.jpg
+
+ArduPilot's Unit Test suite allows for the creation of repeatable tests
+which help prevent regressions in ArduPilot's internal functions.  It is based
+on Google's `GTest Framework <https://https://github.com/google/googletest>`.
+
+Using ArduPilot's Unit Tests can:
+
+   - make your development process more efficient by reducing time spent repeatedly running the same tests to force the same inputs to a function time and again
+   - allow you to repeatedly replicate bad behaviour in an ArduPilot function, and possibly ship that test to a developer capable of fixing the issue ("test-driven-development")
+   - reduce the chances of a regression in an ArduPilot function's behaviour by locking in tests for that function's behaviour
+
+Overview
+========
+
+The Unit test suite is run on ArduPilot's autotest server on most
+commits to the master branch, but can be run locally to vet software
+changes.  Adding tests is straight-forward and encouraged to prove a
+function's correctness after a change.
+
+
+Running AutoTest
+================
+
+AutoTest requires a an environment in which the ArduPilot executables can be compiled.  The easiest way to obtain this is to create a SITL testing environment.  Use the SITL instructions (:ref:`SITL <using-sitl-for-ardupilot-testing>`) to obtain a valid environment.  It is suggested the ArduPilot Vagrant virtual machine configuration files be used to obtain a working environement.
+
+Invocation
+----------
+
+.. note::
+
+   These tests work well under Linux.  cygwin64 is an unknown country.
+
+The unit tests are compiled using the `waf` build tool:
+
+::
+
+    pbarker@bluebottle:~/rc/ardupilot(master)$ ./waf configure --board=linux --debug
+    Setting top to                           : /home/pbarker/rc/ardupilot 
+    Setting out to                           : /home/pbarker/rc/ardupilot/build 
+    Autoconfiguration                        : enabled 
+    .
+    .
+    .
+    Scripting runtime checks                       : enabled 
+    Checking for program 'rsync'                   : /usr/bin/rsync 
+    'configure' finished successfully (1.085s)
+    pbarker@bluebottle:~/rc/ardupilot(master)$ ./waf tests
+    Waf: Entering directory `/home/pbarker/rc/ardupilot/build/linux'
+    Waf: Leaving directory `/home/pbarker/rc/ardupilot/build/linux'
+    .
+    .
+    .
+    tests/test_polygon                                    204433   7056   1352  212841
+    tests/test_segment_intersection                       193839   5832   1288  200959
+    tests/test_vector2                                    210539   6472   1320  218331
+
+    Build commands will be stored in build/linux/compile_commands.json
+    'tests' finished successfully (2.249s)
+    pbarker@bluebottle:~/rc/ardupilot(master)$  
+
+You can then simply run the test from the build directory:
```

Comment:
or use waf --alltests

Reply:
Maybe?

I didn't know about that one... and I guess that could run things other than the unit tests?